Add stylelint configuration

mrliptontea opened the pull request:
# Sane default configuration for [stylelint](https://github.com/stylelint/stylelint)

See [README.md](https://github.com/mrliptontea/sane-project-defaults/blob/stylelint/stylelint/README.md) for rules summary.

## Rules open for discussion

They all are, but I want to point out the ones I'm not very sure about.

### Rules currently in use

* [x] [`max-nesting-depth`](https://stylelint.io/user-guide/rules/max-nesting-depth/) - currently set to 5.
* [x] [`string-quotes`](https://stylelint.io/user-guide/rules/string-quotes/) - single vs. double.
* [x] [`number-leading-zero`](https://stylelint.io/user-guide/rules/number-leading-zero/)- `.5em` vs. `0.5em`.
* [x] [`scss/at-else-if-parentheses-space-before`](https://github.com/kristerkari/stylelint-scss/blob/master/src/rules/at-function-parentheses-space-before/README.md) and [`scss/at-mixin-parentheses-space-before`](https://github.com/kristerkari/stylelint-scss/blob/master/src/rules/at-mixin-parentheses-space-before/README.md) - `function foo ($var)` vs `function foo($var)`.

### Rules not in use, but may be added upon agreement

* [x] [`font-weight-notation`](https://stylelint.io/user-guide/rules/font-weight-notation/)- `font-weight: 700` vs `font-weight: bold`.
* [ ] [`value-border-zero`](https://www.npmjs.com/package/stylelint-value-border-zero) - `border: 0` is shorter while `border: none` doesn't quite work in older versions of IE. And that's another plugin.
* [x] Variable, mixin and function name format - Personally prefer them lowercased any hyphenated, which is consistent with CSS naming.
* [x] Selector format - ids, classes, and elements are not declared in CSS, they're just referenced, therefore checking their format is out of its scope.

Commit:
Add stylelint configuration

Commit:
Add newline at the end of file

Mezzle commented:
- `string-quotes` - I prefer single - what's your reasoning here?
- `number-leading-zero` - I prefer to have the leading 0 here.
- `font-weight-notation` - I'd say use numeric, to make sure we're not hitting inconsistency between browsers/overridden stuff in an OS
- `value-border-zero` - I'm not massively fussed on this - though would prefer to be consistent
- Agree with your comment re: naming
- Agree re selectors - although - I'd really like to say that this should be married up to HTML too. These are defaults, can always be changed per-project if need be.  So, requiring a sane selector name I think seems like it's possibly worthwhile.

mrliptontea commented:
> `string-quotes` - I prefer single - what's your reasoning here?

I prefer single, although I think it's a little weird when using an attribute selector, e.g. `input[type='email']`.

> `number-leading-zero` - I prefer to have the leading 0 here.

Agreed.

> `font-weight-notation` - I'd say use numeric, to make sure we're not hitting inconsistency between browsers/overridden stuff in an OS

Makes sense, and then there's a single convention across all of CSS.

> `value-border-zero` - I'm not massively fussed on this - though would prefer to be consistent

I'll add the plugin then. `border: 0` ensures better compatibility.

> Agree with your comment re: naming

Will add a rule and a regex pattern.

> Agree re selectors - although - I'd really like to say that this should be married up to HTML too. These are defaults, can always be changed per-project if need be. So, requiring a sane selector name I think seems like it's possibly worthwhile.

Hyphenated BEM? E.g. `some-block__sub-element--modifier`

Mezzle commented:
Hmm, I agree it's a little weird. I can't think of why the usual reasoning behind single quotes would be needed here (variables) - can you foresee any issue with double quotes?  Is it possible to enforce single, but double in a selector?

mrliptontea commented:
No, it's either single or double, with an option to avoid escaping.
https://stylelint.io/user-guide/rules/string-quotes/

Other uses for quotes in CSS:
* `content: ""`
* `quotes: "\00ab" "\00bb"`
* `font-family: "Segoe UI", Helvetica, Arial, sans-serif`
* `background-image: url("/img/kitten.png");`

I'd say except for `url` double quotes seem more natural in these cases.

Mezzle commented:
I'm good with double quotes then.  My natural instinct was to jump to single, but double does make sense.

mrliptontea committed:
Use leading zeros

mrliptontea committed:
Use numeric font-weight notation

mrliptontea committed:
Add patterns for selectors, variables, mixins, placeholders, and functions

mrliptontea committed:
Update README.md

mrliptontea commented:
I have added patterns, and updated font-weight notation and leading zeros rules as agreed.
See https://regexr.com/3hv94 for a test.

The `value-border-zero` cannot be used at the moment due to [compatibility issue with latest stylelint](https://github.com/timothyneiljohnson/stylelint-value-border-zero/pull/3), also plugin doesn't seem to be maintained.
